modules/navier_stokes/include/utils/ArrayComponentFunctor.h

```diff
@@ -0,0 +1,76 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseFunctor.h"
+
+/**
+ * This is essentially a forwarding functor that forwards the spatial and temporal evaluation
+ * arguments to the parent array functor and then returns the result indexed at a given component.
+ */
+template <typename T, typename ArrayTypeFunctor>
+class ArrayComponentFunctor : public Moose::FunctorBase<T>
+{
+public:
+  using typename Moose::FunctorBase<T>::ValueType;
+  using typename Moose::FunctorBase<T>::GradientType;
+  using typename Moose::FunctorBase<T>::DotType;
+
+  ArrayComponentFunctor(const ArrayTypeFunctor & array, const unsigned int component)
+    : _array(array), _component(component)
+  {
+  }
+
+  std::pair<bool, const Elem *> isExtrapolatedBoundaryFace(const FaceInfo & fi) const override
+  {
+    return _array.isExtrapolatedBoundaryFace(fi);
+  }
+
+private:
+  /// The parent array functor
+  const ArrayTypeFunctor & _array;
+
+  /// The component at which we'll index the parent array functor evaluation result
+  const unsigned int _component;
+
+  ValueType evaluate(const Moose::ElemArg & elem, const unsigned int state) const override final
+  {
+    return _array(elem, state)[_component];
```

Comment:
Would it result in too long code if we simply called velocity(elem, state)[component] every time instead of creating a component functor for this? Or is this guy doing other things besides that too?

Reply:
This like `greenGaussGradient` only work with functors that have scalar return types. So you cannot simply pass an array-functor into `greenGaussGradient`. `greenGaussGradient` is the reason for the existence of `VectorComponentFunctor` and `ArrayComponentFunctor`. In practice these component-functors are hardly used at run-time but they are necessary for being able to compile. If at some point somebody wants to do an overload of `greenGaussGradient` that works directly with multi-component functors, then we can remove these component functors.